File: Excel/637 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/637 Insert Dash At Non Consecutive Character.xlsx"
input = read_excel(path, range = "A1:A8")
test  = read_excel(path, range = "B1:B8")

process_string = function(string) {
  string %>%
    str_split("") %>%
    unlist() %>%
    {tibble(char = ., value = ifelse(is.na(as.numeric(.)), match(., LETTERS), as.numeric(.)))} %>%
    mutate(dash = ifelse(value - lag(value) != 1, "-", "")) %>%
    replace_na(list(dash = "")) %>%
    unite("char", c("dash", "char"), sep = "") %>%
    pull(char) %>%
    paste0(collapse = "")
}

result = input %>%
  mutate(processed = map_chr(String, process_string))

print(result$processed == test$`Answer Expected`)
# [1]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE 
# one AB shoud be a pair one time in last string

File: Excel/600-699/637/637 Challenge.py
import pandas as pd
import numpy as np

path = "637 Insert Dash At Non Consecutive Character.xlsx"
input = pd.read_excel(path, usecols="A", nrows=8)
test = pd.read_excel(path, usecols="B", nrows=8)

def process_string(string):
    result = [string[0]]
    for i in range(1, len(string)):
        if (ord(string[i]) - ord(string[i-1]) != 1):
            result.append('-')
        result.append(string[i])
    return ''.join(result)

input['processed'] = input.iloc[:, 0].apply(process_string)

print(test['Answer Expected'] == input['processed'])

# 0     True
# 1     True
# 2     True
# 3     True
# 4     True
# 5     True
# 6    False  AB in this string can be pair once. 
